03-rnaseq/clustering/clustering_heatmap_example.Rmd

```diff
@@ -0,0 +1,220 @@
+
+```{r Install and attach the pheatmap and DESeq2 libraries}
+if (!("pheatmap" %in% installed.packages())) {
+  # Install pheatmap
+  install.packages("pheatmap", update = FALSE)
+}
+
+if (!("DESeq2" %in% installed.packages())) {
+  # Install DESeq2
+  BiocManager::install("DESeq2", update = FALSE)
+}
+```
+
+Attach the `pheatmap` and `DESeq2` libraries:
+
+```{r}
+# Attach the `pheatmap` library
+library(pheatmap)
+
+# Attach the `DESeq2` library
+library(DESeq2)
+
+# Magrittr pipe
+`%>%` <- dplyr::`%>%`
+
+# Set the seed so our results are reproducible:
+set.seed(12345)
+```
+
+Create output folder.
+
+```{r}
+# Define the file path to the plots directory
+plots_dir <- "plots" # Replace with path to desired output plots directory
+
+# Create the plots folder if it doesn't exist
+if (!dir.exists(plots_dir)) {
+  dir.create(plots_dir)
+}
+```
+
+## 2) Import and set up data
+Data downloaded from refine.bio include a metadata tab separated values ("tsv")
+file and a data tsv file. This chunk of code will read the both tsv files and 
+add them as data.frame objects to your environment.
+
+The data that we downloaded from refine.bio for this analysis is a dataset, with
+a total of four samples, containing RNA-sequencing results for mouse primary
+natural killer (NK) cells and NK cells treated with DKK2 for 24 hours.
+
+You can download this dataset by going to the following link and clicking 
+`Download Now`: https://www.refine.bio/experiments/SRP110235/mouse-natural-killer-cells-response-to-dkk2-treatment
+
+```{r Import data from .tsv files}
+# Read in metadata tsv file
+metadata <- readr::read_tsv(file.path(
+  "data", # Replace with path to your metadata file
+  "metadata_SRP110235.tsv" # Replace with the name of your metadata file
+)) %>%
+  dplyr::select(refinebio_accession_code, refinebio_title)
+
+# Read in data tsv file
+df <- readr::read_tsv(file.path(
+  "data", # Replace with path to your data file
+  "SRP110235.tsv" # Replace with the name of your data file
+)) %>%
+  tibble::column_to_rownames("Gene")
+```
+
+Let's ensure that the metadata and data are in the same sample order. 
+
+```{r}
+# Make the data in the order of the metadata
+df <- df %>% dplyr::select(metadata$refinebio_accession_code)
+
+# Check if this is in the same order
+all.equal(colnames(df), metadata$refinebio_accession_code)
+```
+
+## 3) Choose genes of interest
+Although you may want to create a heatmap including all of the genes in the set,
+alternatively, the heatmap could be created using only genes of interest. 
+For this example, we will sort genes by variance, but there are many alternative
+criterion by which you may want to sort your genes <i>eg</i> fold change,
+t-statistic, membership to a particular gene ontology, so on. 
+
+```{r Choose genes}
+# Calculate the variance for each gene
+variances <- apply(df, 1, var)
```

Comment:
I think it's a good idea 👍 (we do use the `apply` function elsewhere)

Reply:
This is the only comment I have yet to address because I have not yet thought of an appropriate tidyverse strategy to replace what the apply function is doing here.

Reply:
Apply is fine here. This is essentially a matrix, and you are doing a row-wise calculation, which tidyverse is less adept at. 
A suggestion you could make is to use `matrixStats::rowVars()`, which is much faster than `apply()`, but I think apply is fine.